### PR TITLE
AK+Kernel: Make IntrusiveRedBlackTree capable of holding non-raw pointers and use it for the cr3 map

### PR DESCRIPTION
**Pull request by IdanHo**
Opening this PR so I can get some feedback on the `IntrusiveRedBlackTree` changes I made before I start converting more `HashMap` users to it.

### CODE REVIEW

File: Kernel/Memory/PageDirectory.h [hunk 1]
```diff
@@ -46,6 +47,9 @@ class PageDirectory : public RefCounted<PageDirectory> {
 
     RecursiveSpinlock& get_lock() { return m_lock; }
 
+    // This has to be public to allow for the global singleton
```
[the remark this answers is not in the document]
Comment (IdanHo) [reply Kernel/Memory/PageDirectory.h]: oops, fixed!

**Review by alimpfard — APPROVED** [review]
LGTM!

**Comment by bgianfo** [comment]
Clang doesn't seem as forgiving of syntax:
```
/home/vsts/work/1/s/./AK/IntrusiveRedBlackTree.h:55:21: error: missing 'typename' prior to dependent type name 'BaseTree::Node'
        static_cast<BaseTree::Node&>(node).key = key;
                    ^~~~~~~~~~~~~~
                    typename 
1 error generated.
```

**Comment by IdanHo** [comment]
> Clang doesn't seem as forgiving of syntax:
> 
> ```
> /home/vsts/work/1/s/./AK/IntrusiveRedBlackTree.h:55:21: error: missing 'typename' prior to dependent type name 'BaseTree::Node'
>         static_cast<BaseTree::Node&>(node).key = key;
>                     ^~~~~~~~~~~~~~
>                     typename 
> 1 error generated.
> ```

Whoops, fixed!

**Comment by awesomekling** [comment]
"the storing cr3 mappings" -> "storing the cr3 mappings"

**Comment by IdanHo** [comment]
> "the storing cr3 mappings" -> "storing the cr3 mappings"

Fixed!

**Pull request merged**